src/UniqueMsg.chpl

```diff
@@ -48,7 +48,64 @@ module UniqueMsg
         segments.a = (+ scan counts) - counts;
         return (permutation, segments);
       }
+      if hasStr && n == 1 {
+        // Only one array which is a strings
+        var (myNames, _) = namesList[0].splitMsgToTuple('+', 2);
+        var strings = getSegString(myNames, st);
+        const lens = strings.getLengths();
+        var max_bytes = max reduce lens;
+        if max_bytes < 16 {
+          // bytes contained in strings < 128 bits, so concatenating is better than the hash
+          ref oa = strings.offsets.a;
+          ref va = strings.values.a;
+          if max_bytes < 8 {
+            // we only need one uint array
+            var numeric = makeDistArray(strings.offsets.size, uint);
+            forall (num, o, l) in zip(numeric, oa, lens) {
+              var concat: uint;
+              for v in va[o..#l] {
+                concat = concat<<8 + v:uint;
+              }
+              num = concat;
+            }
+            const concatName = st.nextName();
+            st.addEntry(concatName, new shared SymEntry(numeric));
+            names = [concatName];
+            types = ["pdarray"];
+            hasStr = false;
+            // we can now use the !hasStr logic below
+          }
+          else {
+            // we need two uint arrays
+            var numeric1 = makeDistArray(strings.offsets.size, uint);
+            var numeric2 = makeDistArray(strings.offsets.size, uint);
+            forall (num1, num2, o, l) in zip(numeric1, numeric2, oa, lens) {
+              var concat: uint;
+              var half = (l/2):int;
+              for v in va[o..#half] {
+                concat = concat<<8 + v:uint;
+              }
+              num1 = concat;
 
+              concat = 0;
+              for v in va[(o+half)..#half] {
+                concat = concat<<8 + v:uint;
+              }
+              num2 = concat;
+            }
+            // chapel doesn't like changing the size of names, so copy !hasStr logic
+            const concat1Name = st.nextName();
+            const concat2Name = st.nextName();
+            st.addEntry(concat1Name, new shared SymEntry(numeric1));
+            st.addEntry(concat2Name, new shared SymEntry(numeric2));
+            var (totalDigits, bitWidths, negs) = getNumDigitsNumericArrays(names, st);
+            if totalDigits <= 2 { return skipSortHelper(2*uint(bitsPerDigit), mergeNumericArrays(2, size, totalDigits, bitWidths, negs, names, st)); }
+            if totalDigits <= 4 { return skipSortHelper(4*uint(bitsPerDigit), mergeNumericArrays(4, size, totalDigits, bitWidths, negs, names, st)); }
+            if totalDigits <= 6 { return skipSortHelper(6*uint(bitsPerDigit), mergeNumericArrays(6, size, totalDigits, bitWidths, negs, names, st)); }
+            if totalDigits <= 8 { return skipSortHelper(8*uint(bitsPerDigit), mergeNumericArrays(8, size, totalDigits, bitWidths, negs, names, st)); }
+          }
+        }
+      }
```

Comment:
Could we pull this logic out into a function since it is reused below to reduce duplicated code?

Reply:
You're exactly right! It's duplicated in the `assumeSortedShortcut` and `uniqueAndCount` functions. Both of these functions do basically the same thing (except `assumeSortedShortcut` skips the sort). I implemented these for an important bug a while back so I went with the approach with the quickest turn around. I am planning on combining the two functions into one during the next iteration which will eliminate the duplicated code